### PR TITLE
[core][autoscaler][v2] Reconciler - handle out-of bands cloud instances [12/x] 

### PR DESCRIPTION
**Pull request by rickyyx**
## Why are these changes needed?
There are cases where cloud provider might have launched cloud nodes, which Autoscaler wasn't aware yet, i.e. the cluster has already scaled up more than the IM instances. When this happens, instead of shutting them down, we will create IM instances with ALLOCATED status for them. As of now, there are 3 cases we already know that this could happen: 
1. Head node: head node is always started before autoscaler for now, so when Autoscaler starts, it will see an unmanaged already running cloud node, we will initialize the cloud node as ALLOCATED. 
2. K8s scaled up worker nodes: KubeRay would launch new worker nodes if users changed `min_workers` value in the k8s definitions, and those workers nodes would be started by Kuberay's cloud providers. 
3. Local cluster mode, where ray nodes might get started by users and we still need the autoscaler monitor to work. 

This behaviour is ok since:
- If those "extra" cloud nodes are due to cases 1, 2 above, they would transition to RAY_RUNNING as a normal IM instances, it's just they were not started by IM (from QUEUED statuses)
- else those "extra" cloud nodes are "leaked": we will have the stuck instances reconciliation mechanism + idle nodes termination to shut them down, we also have the full knowledge that these nodes were started by the cloud providers so we could later on let users know. 

Either way, upserting new instances with `ALLCOATED` status instead of shutting them down for extra cloud instances seems to be ok. 

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/autoscaler/v2/instance_manager/reconciler.py [hunk 1]
```diff
@@ -106,6 +107,52 @@ def reconcile(
...
+        assert head_cloud_instance, (
+            "Head node is not found. Make sure the head node "
+            "is started with RAY_CLOUD_INSTANCE_ID set, and it has the needed labels "
+            "specified in ray/autoscaler/tags.py ."
+        )
+
+        # Initialize the head node for the instance manager.
+        updates = {}
+        head_node_instance_id = InstanceUtil.random_instance_id()
+        updates[head_node_instance_id] = IMInstanceUpdateEvent(
```
Comment (jjyao) [inline python/ray/autoscaler/v2/instance_manager/reconciler.py:143]: So this will create and update the status to ALLOCATED in one go?

Comment (rickyyx) [reply python/ray/autoscaler/v2/instance_manager/reconciler.py:143]: Yes - see this PR for the instance manager change for upsert: https://github.com/ray-project/ray/pull/43098

Comment (jjyao) [reply python/ray/autoscaler/v2/instance_manager/reconciler.py:287]: Explicitly set upsert=true for cases where upsert is expected.

File: python/ray/autoscaler/v2/instance_manager/config.py [hunk 2]
```diff
@@ -299,20 +300,32 @@ def get_node_type_configs(self) -> Dict[NodeType, NodeTypeConfig]:
...
                 min_worker_nodes=node_config.get("min_workers", 0),
-                max_worker_nodes=node_config.get("max_workers", 0),
+                max_worker_nodes=max_workers_nodes,
                 resources=node_config.get("resources", {}),
                 labels=node_config.get("labels", {}),
                 launch_config_hash=launch_config_hash,
             )
         return node_type_configs
 
+    def get_head_node_type(self) -> NodeType:
```
Comment (jjyao) [inline python/ray/autoscaler/v2/instance_manager/config.py:323]: Is this the v1 logic?

File: python/ray/autoscaler/v2/instance_manager/instance_manager.py [hunk 3]
```diff
@@ -199,13 +208,20 @@ def _apply_update(instance: Instance, update: InstanceUpdateEvent):
             assert update.instance_type, "REQUESTED update must have instance_type"
             instance.launch_request_id = update.launch_request_id
             instance.instance_type = update.instance_type
+        elif update.new_instance_status == Instance.TERMINATING:
+            assert (
+                update.cloud_instance_id
+            ), "TERMINATING update must have cloud instance id"
+            pass
```
Comment (jjyao) [inline python/ray/autoscaler/v2/instance_manager/instance_manager.py:215]: nit: not needed

File: python/ray/autoscaler/v2/scheduler.py [hunk 4]
```diff
@@ -1024,9 +1024,17 @@ def _select_nodes_to_terminate(
...
+        for i, node in enumerate(nodes):
+            if node.node_kind == NodeKind.HEAD:
+                # Remove the head node from the list.
+                head_node = nodes.pop(i)
+                break
+
         terminated_nodes, remained_nodes = (
             nodes[:num_to_terminate],
-            nodes[num_to_terminate:],
+            nodes[num_to_terminate:] + ([head_node] if head_node else []),
```
Comment (jjyao) [inline python/ray/autoscaler/v2/scheduler.py:1037]: can head_node be None?

File: python/ray/autoscaler/v2/instance_manager/reconciler.py [hunk 5]
```diff
@@ -760,15 +782,12 @@ def _handle_ray_status_transition(
...
-            if cloud_instance_id not in im_instances_by_cloud_instance_id:
-                # This is a ray node that's not managed by the instance manager.
-                # or we haven't discovered the instance yet. There's nothing
-                # much we could do here.
-                logger.info(
-                    f"Ray node {binary_to_hex(ray_node.node_id)} has no matching "
-                    f"instance with cloud instance id={cloud_instance_id}."
-                )
-                continue
+            assert cloud_instance_id in im_instances_by_cloud_instance_id, (
```
Comment (jjyao) [inline python/ray/autoscaler/v2/instance_manager/reconciler.py:785]: For out-of-band ray node, we do nothing?

File: python/ray/autoscaler/v2/instance_manager/reconciler.py [hunk 6]
```diff
@@ -1113,6 +1136,29 @@ def _warn_stuck_instances(
...
+        Check if the head node is running and ready.
+
+        If we scale up the cluster before head node is running,
+        it would cause issues when launching the worker nodes.
+
+        Args:
+            instance_manager: The instance manager to reconcile.
+
+        Returns:
+            True if the head node is running and ready, False otherwise.
```
Comment (jjyao) [inline python/ray/autoscaler/v2/instance_manager/reconciler.py:1151]: when it can be false?

File: python/ray/autoscaler/v2/instance_manager/reconciler.py [hunk 7]
```diff
@@ -1321,6 +1382,97 @@ def _install_ray(
 
         Reconciler._update_instance_manager(instance_manager, version, updates)
 
+    @staticmethod
+    def _fill_autoscaling_state(
```
Comment (jjyao) [inline python/ray/autoscaler/v2/instance_manager/reconciler.py:1386]: unrelated?

File: python/ray/autoscaler/v2/instance_manager/reconciler.py (hunk 6 above)
Comment (jjyao) [reply python/ray/autoscaler/v2/instance_manager/reconciler.py:1151]: comment

File: python/ray/autoscaler/v2/instance_manager/reconciler.py (hunk 5 above)
Comment (rickyyx) [reply python/ray/autoscaler/v2/instance_manager/reconciler.py:785]: we will upsert out-of-band ray node's cloud instance with a new ALLOCATED IM instance just like how we handle out-of-band cloud instance. And let the reconcilation handle the rest.

File: python/ray/autoscaler/v2/instance_manager/reconciler.py [hunk 8]
```diff
@@ -1197,13 +1239,21 @@ def _scale_cluster(
             reply.infeasible_cluster_resource_constraints
         )
 
+        if not Reconciler._is_head_node_running(instance_manager):
+            # We shouldn't be scaling the cluster until the head node is ready.
```
Comment (jjyao) [inline python/ray/autoscaler/v2/instance_manager/reconciler.py:1243]: explain why

**Review by jjyao — APPROVED** [review]

File: python/ray/autoscaler/v2/instance_manager/config.py (hunk 2 above)
Comment (rickyyx) [reply python/ray/autoscaler/v2/instance_manager/config.py:323]: yes - commented on the behaviour

File: python/ray/autoscaler/v2/scheduler.py (hunk 4 above)
Comment (rickyyx) [reply python/ray/autoscaler/v2/scheduler.py:1039]: yes - commented, i.e. when the head node was just allocated, but not yet reported ray running.
